### ISSUE TITLE
Distinguish between "public" and "private" dependencies

### ISSUE THREAD

**Issue by johan-boule** [jpakkane/meson#106]
The dependency system in meson should make it possible to specify if a dependency is used only as an implementation detail (say, a "private" dep), or if it's exposed transitively to clients with #include in public headers (say a "public" dep).
The auto-generated pkg-config file should reflect that using appropriate Requires and Requires.private.
Note: I however think it would add unnecessary complexity to include a third kind of dependency like CMake does. Also, pkg-config choosed a middle-ground where CFlags are always transitive (i.e. Requires.private does expose CFlags transitively; only Libs and Libs.private are not exposed through Requires.private), so it would be a bit pointless to go beyond these two types of dependencies provided by pkg-config.

**Comment by johan-boule** [comment]
"only Libs and Libs.private are not exposed through Requires.private"
IIRC this is true unless one passes the --static option to pkg-config, but that's only used when linking static programs.

**Comment by jpakkane** [comment]
The pkg-config generator in Meson is only ever meant to be used for simple pkg-config files. More complex files need to be generated with configure_file(). The reason for this is that there is no simple way of telling what should be private deps and which should not. The Meson configuration language would need to add the option to add this information, which would be a lot of code to maintain and would be _more_ complex than having users generate pc files themselves.

**Comment by johan-boule** [comment]
It sure adds complexity to meson itself, but from the user perspective, I don't think it wouldn't change much things. For example, it could simply be a new private_dependencies keyword, e.g. :
```
executable('manydeps', 'file.c', dependencies : [dep1, dep2], private_dependencies : [dep3, dep4])
```

**Comment by ignatenkobrain** [comment]
Because it's not needed much code changes I think it can be implemented

**Comment by pwithnall** [comment]
Knowing the set of public dependencies is also useful when running external tools on the code, such as gtk-doc, hotdoc, or the Clang static analyser, which all need to know the CFLAGS and CPPFLAGS (in autotools terms) of the public dependencies of a project; but don't need to know them of the private/internal dependencies.

**Comment by sarum9in** [comment]
This kinda looks similar to the proposal I made in #3128 but if we implement this, we could just reuse core language dependency object in pkgconfig better.

**Comment by Ericson2314** [comment]
This is a good idea because it isn't just needed by cmake. If I have an executable that depends on a library which a some `static inline` function, I need to get the transitive public deps for linking. This can all be internal to a meson project.

**Comment by Ericson2314** [comment]
Having had to write some more CMake recently, I'm more convinced this really is a good idea.

**Comment by kasper93** [comment]
Similar issue affects libplacebo: https://code.videolan.org/videolan/libplacebo/-/merge_requests/628

**Comment by Ericson2314** [comment]
Getting public vs private right for https://github.com/nixos/nix has definitely taken a bit more cognitive load than I would have liked. This has gotten me thinking about how the feature would be implemented in a Meson-ish way. Here's what I came up with:

1. Executables and Libraries only have one kind of dep, as today. This is because there is no public vs private distinction for machine code. (Technically, we could distinguishing what is in the ABI or not, but that is not how CMake uses its public vs private distinction. That is future work.)

2. headers should have dependencies. Rather than just having the (rather ugly and imperative) `install_headers` as a verb, we should have a `headers` object with `install: bool`.

4. `declare_dependency` should not just be aware of executables and libraries, but also be aware of header objects. *here* is where the public and private distinction should be made: executable/library deps are *private* deps of the declared dependency, header deps are *public* deps.

5. The `pkg-config` module should, instead of working from executables and libraries, work from declared dependencies. Then, and only then, do they have all the information they need to do get the public vs private fields to be correct.

**Comment by eli-schwartz** [comment]
> 4. The `pkg-config` module should, instead of working from executables and libraries, work from declared dependencies. Then, and only then, do they have all the information they need to do get the public vs private fields to be correct.

This echoes thoughts I've had myself (well, really that one should be defined in terms of the other although which direction it goes isn't crucial in my mind).



> 2. headers should have dependencies. Rather than just having the (rather ugly and imperative) `install_headers` as a verb, we should have a `headers` object with `install: bool`.
> 
> 3. `declare_dependency` should not just be aware of executables and libraries, but also be aware of header objects. _here_ is where the public and private distinction should be made: executable/library deps are _private_ deps of the declared dependency, header deps are _public_ deps.


I don't really understand what you mean by this. What would a dependency on a header indicate?

I'm exclusively aware of this as a concept, for the bazel build system. My understanding is they define "header objects are a dependency" to mean "the build system mounts a fuse filesystem that filters the files visible in a directory so that adding -I paths on your headers only allows the compiler to see the headers that have been described as permitted for that target".


It doesn't solve the problem of using the correct -I paths in accordance with your nested header directory hierarchy. The (since 0.63.0) feature `preserve_path: true` does, in a sense, but only inasmuch as it allows you to tag the fact that a group of headers installed as a collection of headers via `install_headers()`, should map to an -I argument pointing at the directory the `install_headers()` invocation was performed in. It doesn't really add new capabilities.

Am I missing something here?

**Comment by Ericson2314** [comment]
Sorry @eli-schwartz, I was missing up "headers" and "include directories". Everything I said that should go on the headers should instead go on the include directories.

(I do wish we could unify headers and include directories a bit: requiring FUSE is a bit to much, but just making sure the list of headers and the list of include directories aligned would be more practical, but this is an orthgonal issue.)

Basically, I see a "public C/C++ interface" as a 3 part concept:

1. Some headers to be installed
2. Some include directories, which match where those headers exist (in the src tree or in the installation dir)
3. Some dependencies, which provide other `-I` flags necessary to use these headers. (And other CFLAGS/CXXFLAGS too, within reason.)

**Comment by kasper93** [comment]
> Libraries only have one kind of dep, as today. This is because there is no public vs private distinction for machine code.

I don't agree with this. Libraries can have public include directories (headers) and be used only as private dependency to build another library, which shouldn't export those public include directories.

In CMake it is done by `target_link_libraries(<target> <PRIVATE|PUBLIC|INTERFACE> <item>`. In Meson it would be to have already proposed `as_private()` function, that would convert library dep to private. But ultimately, there has to be a way to distinguish those two cases.

**Comment by Ericson2314** [comment]
@kasper93 Please read what I wrote more closely. I am saying that since libraries are *not* responsible for installing headers --- a fact true today, they are not responsible for the dependencies of those headers.

There is no missing expressive power. When just the library but not its headers/include dirs depends on something, that is a private dependency and no `-I` flags will be propogated downstream. (And per `Requires.private` `-l` flags will also only be propogated downstream for static linking.)

**Comment by xclaesse** [comment]
An important read about the 3 types of dependencies can be found here: https://bugs.freedesktop.org/show_bug.cgi?id=105572#c1.

I wrote a patch to add `Requires.internal` to pkg-config 3 years ago: https://gitlab.freedesktop.org/pkg-config/pkg-config/-/merge_requests/8. IIRC that got merged in pkgconf.

I've always wanted in Meson to _somehow_ merge `declare_dependency()` and `pkg.generate()` because they share a lot of info.

**Comment by Ericson2314** [comment]
@xclaesse Is there a valid use-case for `Requires.private` with that, or does  `Requires.internal` replace it for all regular use-cases? (I suppose there could be some "internal use only headers", but IMO that would be better served by a separate pc file.)

> Basically declare_dependency, pkg.generate and meson.override_dependency should all be a single function.

Yes

> But I'm not sure how exactly and what would be the test API. 

I think that is where thinking about headers/include directories as proper target can help. :) C/C++ interface is a concept, and it handles public CFLAGS/CXXFLAGS. And then remaining clare_dependency, pkg.generate and meson.override_dependency *just* handles what things get installed and aligning libraries and their corresponding headers/include dirs/flags.

**Comment by xclaesse** [comment]
> Is there a valid use-case for `Requires.private` with that, or does `Requires.internal` replace it for all regular use-cases?

Yes, there are valid cases for all 3 public/private/internal dependencies. See the fairly long description from @smcv in the bugzilla link.

**Comment by Ericson2314** [comment]
OK Yeah I don't love the "Critical dependencies" vs "Visible dependencies" distinction because it's not precise, suffering from https://en.wikipedia.org/wiki/Sorites_paradox ("How often does something need to be `#include`d before it is critical?)

(I think it is better to have separate PC files for different features. You want a bonus thing? Add another requires, and get additional `-I` and `-D` flags, at the cost of additional deps (which become additional `-l` when static linking.)

Regardless though, we can support this notion by having the "critical" vs "visible" distinction on the `headers`/`include_directories` target. The insight here is the library (i.e. machine code) doesn't care, this is just a header-side distinction --- variations within the "public deps" category.

**Comment by eli-schwartz** [comment]
> OK Yeah I don't love the "Critical dependencies" vs "Visible dependencies" distinction because it's not precise, suffering from https://en.wikipedia.org/wiki/Sorites_paradox ("How often does something need to be `#include`d before it is critical?)

It's a bad distinction that doesn't clearly associate with the defined purposes of the three fields.

A pkg-config:
- Requires is a public dependency, if you use foo's headers then foo's headers also include bar's headers and expose bar's types to your code, so you need to compile *and* link against bar, because ***foo publicizes bar in its API/ABI***.
- Requires.internal is a private dependency, if you use foo's headers then foo is defined in terms of foo and foo alone. But foo's library internally uses bar, so you have to link to bar for static linkage *only*. In this case, ***foo totally hides bar, but static link still needs to work***.
- Requires.private is a sort-of-private, sort-of-public dependency, if you use foo's headers then foo is implemented in terms of bar, but safely wraps it so your object code only depends on foo. ***foo publicizes bar in its API alone (and static link still needs to work)***.


"Critical" is a bad term not because of philosophical paradoxes, nor because of "how much do you care". It's a bad term because you either need something or you don't, and "you need both" is not really the same thing as "you need it to the maximum extent of need", which is what criticality implies.

**Comment by Ericson2314** [comment]
Ah yes. I can see `Requires.private` being abused for all sorts of optional transitive implementation-in-headers dependencies shenanagins (templates,`static inline` functions). Yuck!

**Comment by Ericson2314** [comment]
BTW, what I am doing in https://github.com/NixOS/nix/pull/10973 to protoype this to dedup a bunch of subprojects may be of interest.
